data/cube/impl/BaseRow.js

```diff
@@ -0,0 +1,150 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+
+import {has, isEmpty, reduce} from 'lodash';
+
+
+/**
+ * Base class for a view row.
+ */
+export class BaseRow {
+
+    view = null;
+    id = null;
+    parent = null;
+    children = null;
+    data;
+
+    get isLeaf()        {return false}
+    get isAggregate()   {return false}
+    get isBucket()      {return false}
+
+    constructor(view, id) {
+        this.view = view;
+        this.id = id;
+        this.data = {id, _meta: this};
+    }
+
+    //-----------------------
+    // For all rows types
+    //------------------------
+    noteBucketed(bucketSpec, bucketVal) {
+        this.data.buckets = this.data.buckets ?? {};
+        this.data.buckets[bucketSpec.name] = bucketVal;
+        this.children?.forEach(it => it.noteBucketed(bucketSpec, bucketVal));
+    }
+
+    // Determine what should be exposed as the actual children in the
+    // row data.  This where we lock, skip degenerate rows, etc.
+    applyVisibleChildren() {
+        let {children, view, data} = this,
+            {lockFn, omitFn} = view.cube;
+
+        if (!children) return null;
+
+        // Remove all children from the data if the query is not configure to include leaves and
+        // this row has leaves as children
+        if (!view.query.includeLeaves && children[0]?.isLeaf) {
```

Comment:
I wonder if we have a potential bug here. Say we are bucketing leaf rows but `includeLeaves` is `false` in the query - and the first child here is a bucket row containing only leaves. In that case we would not remove any of those leaves, right? Should this be a `chilldren.some(it => it.isLeaf)` ?

Reply:
we resolved -- we don't create buckets of leaves